plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go

```diff
@@ -0,0 +1,380 @@
+	wg  sync.WaitGroup
+}
+
+// Start the Cisco MDT service
+func (c *CiscoTelemetryMDT) Start(acc telegraf.Accumulator) error {
+	var err error
+	c.acc = acc
+	c.listener, err = net.Listen("tcp", c.ServiceAddress)
+	if err != nil {
+		return err
+	}
+
+	switch c.Transport {
+	case "tcp":
+		// TCP dialout server accept routine
+		c.wg.Add(1)
+		go func() {
+			c.acceptTCPClients()
+			c.wg.Done()
+		}()
+
+	case "grpc":
+		var opts []grpc.ServerOption
+
+		if c.EnableTLS {
+			tlsConfig, err := c.ServerConfig.TLSConfig()
+			if err != nil {
+				return err
+			}
+
+			opts = append(opts, grpc.Creds(credentials.NewTLS(tlsConfig)))
+		}
+
+		if c.MaxMsgSize > 0 {
+			opts = append(opts, grpc.MaxRecvMsgSize(c.MaxMsgSize))
+		}
+
+		c.grpcServer = grpc.NewServer(opts...)
+		dialout.RegisterGRPCMdtDialoutServer(c.grpcServer, c)
+
+		c.wg.Add(1)
+		go func() {
+			c.grpcServer.Serve(c.listener)
+			c.wg.Done()
+		}()
+
+	default:
+		c.listener.Close()
+		return fmt.Errorf("invalid Cisco MDT transport: %s", c.Transport)
+	}
+
+	return nil
+}
+
+// AcceptTCPDialoutClients defines the TCP dialout server main routine
+func (c *CiscoTelemetryMDT) acceptTCPClients() {
+	// Keep track of all active connections, so we can close them if necessary
+	var mutex sync.Mutex
+	clients := make(map[net.Conn]struct{})
+
+	for {
+		conn, err := c.listener.Accept()
+		if neterr, ok := err.(*net.OpError); ok && (neterr.Timeout() || neterr.Temporary()) {
+			continue
+		} else if err != nil {
+			break // Stop() will close the connection so Accept() will fail here
+		}
+
+		mutex.Lock()
+		clients[conn] = struct{}{}
+		mutex.Unlock()
+
+		// Individual client connection routine
+		c.wg.Add(1)
+		go func() {
+			log.Printf("D! Accepted Cisco MDT TCP dialout connection from %s", conn.RemoteAddr())
+			if err := c.handleTCPClient(conn); err != nil {
+				c.acc.AddError(err)
+			}
+			log.Printf("D! Closed Cisco MDT TCP dialout connection from %s", conn.RemoteAddr())
+
+			mutex.Lock()
+			delete(clients, conn)
+			mutex.Unlock()
+
+			conn.Close()
+			c.wg.Done()
+		}()
+	}
+
+	// Close all remaining client connections
+	mutex.Lock()
+	for client := range clients {
+		if err := client.Close(); err != nil {
+			log.Printf("E! Failed to close TCP dialout client: %v", err)
+		}
+	}
+	mutex.Unlock()
+}
+
+// Handle a TCP telemetry client
+func (c *CiscoTelemetryMDT) handleTCPClient(conn net.Conn) error {
+	var err error
+
+	// TCP Dialout telemetry framing header
+	var hdr struct {
+		MsgType       uint16
+		MsgEncap      uint16
+		MsgHdrVersion uint16
+		MsgFlags      uint16
+		MsgLen        uint32
+	}
+
+	var payload bytes.Buffer
+
+	for {
+		// Read and validate dialout telemetry header
+		if err = binary.Read(conn, binary.BigEndian, &hdr); err != nil {
+			break
+		}
+
+		maxMsgSize := tcpMaxMsgLen
+		if c.MaxMsgSize > 0 {
+			maxMsgSize = uint32(c.MaxMsgSize)
+		}
+
+		if hdr.MsgLen > maxMsgSize {
+			err = fmt.Errorf("dialout packet too long: %v", hdr.MsgLen)
```

Comment:
`return fmt.Errorf(...`, do this at other break points in this function.

Reply:
Done